### ISSUE TITLE
Proposal-Discussion: Centralize shared model components (MLP, Attention, Norms)

### ISSUE THREAD

**Issue by AvgJoe-cpu** [ZHZisZZ/dllm#35]
Hi, thanks for open-sourcing this project!

**Goals:**
- Enable model variants based on existing architectures (attention mechanisms, MLPs, parameter sizes)
- Standardize model class signatures for better uniformity

**Intuition:** Models share core components (MLP, Attention, RoPE) that could be centralized and reused while maintaining transformers compatibility.

## I. Atomic Ops

Make atomic operations lean by avoiding config objects. Use functional wrappers instead of nn.Modules. We have 2 options: use default args or create separate wrappers for each variant. Since variants are few, we tend towards using separate wrappers.

**1) Dropout**

We avoid an additional check "if p>0" because dropout does this under the hood (returns identity for p==0). We avoid constructing a nn.Module and reduce the signature of the affected modules. (https://github.com/pytorch/pytorch/blob/1d705b4b075e32540293d1717b582442a66ffcce/aten/src/ATen/native/Dropout.cpp#L44)

```python
def _dropout(x, dropout_p, training: bool): 
    return F.dropout(x, p=dropout_p, training=training, inplace=True)
```

**2) Activations**

PyTorch includes functional activations (https://docs.pytorch.org/docs/stable/nn.functional.html). Identifies 400+ variants ([arXiv:2402.09092](https://arxiv.org/abs/2402.09092)), a non-exhaustive list as of 2024.

```python
def _gelu(x): return F.gelu(x)
```

**3) Norms**

Weight and bias are None (can be added for learnable parameters).

```python
def _norm(x): return F.layer_norm(x, normalized_shape=(x.shape[-1],), weight=None, bias=None)

def _rms(x): return F.rms_norm(x, normalized_shape=(x.shape[-1],), weight=None)

def _rms_w(x, weight): return F.rms_norm(x, normalized_shape=(x.shape[-1],), weight=weight) # learnable weight
```

### **1. MLP**

- Create activation registry ([transformers #27846](https://github.com/huggingface/transformers/issues/27846))
- Keep class generic (project up-down pattern)
- Minimal signature without config; use `factory_kwargs` for device/dtype management

Note: Linear layers use nn.Linear (could be parameters). Gated variants may need separate MLPGated module.

**Baseline**

Minimal signature: 3 args plus activation (no config). Activation registry enables scaling.

```python
ACT2FN = {'gelu': _gelu}

class MLP(nn.Module):
  def __init__(self, hidden_dim, exp_factor, dropout_p, act_fn, device=None, dtype=None):
    super().__init__()
    factory_kwargs = {'device': device, 'dtype': dtype}
    
    self.dropout_p = dropout_p
    self.fc1 = nn.Linear(hidden_dim, hidden_dim * exp_factor, bias=False, **factory_kwargs)
    self.fc2 = nn.Linear(hidden_dim * exp_factor, hidden_dim, bias=False, **factory_kwargs)
    self.act_fn = ACT2FN[act_fn]

  def forward(self, x):
    return self.fc2(_dropout(self.act_fn(self.fc1(x)), self.dropout_p, self.training))
```

**Improved handling**

Modern approach using `nn.factory_kwargs()` for cleaner device management ([docs](https://docs.pytorch.org/docs/stable/generated/torch.nn.factory_kwargs.html)). This enables meta device initialization ([docs](https://docs.pytorch.org/docs/stable/meta.html#idioms-for-working-with-meta-tensors)).

```python
class MLP(nn.Module):
    def __init__(self, hidden_dim, exp_factor, dropout_p, act_fn, **kwargs):
        super().__init__()
        factory_kwargs = nn.factory_kwargs(kwargs)
        # as above

    def forward(self, x):
        return self.fc2(_dropout(self.act_fn(self.fc1(x)), self.dropout_p, self.training))
```

Example with meta device:

```python
with torch.device('meta'):
    mlp = MLP(hidden_dim=240, exp_factor=4, dropout_p=0.1, act_fn='gelu')
print(mlp.fc1.weight.device)  # meta
```

### **2. Attention**

- Use SDPA which supports GQA (Grouped Query Attention) by default (Dropout supported as arg - omitted for simplicity)(https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html)
- For attention gating variants (headwise/elementwise non-linearity, sparsity), see [arXiv:2505.06708](https://arxiv.org/abs/2505.06708) - may call for a separate module
- FA import unnecessary unless using custom implementation ([Triton tutorial](https://triton-lang.org/main/getting-started/tutorials/06-fused-attention.html#sphx-glr-getting-started-tutorials-06-fused-attention-py))

```python
def _attention(q, k, v):
  return F.scaled_dot_product_attention(query=q, key=k, value=v, is_causal=False, enable_gqa=False)
```
```python
def _attention_gq(q, k, v):
  return F.scaled_dot_product_attention(query=q, key=k, value=v, is_causal=False, enable_gqa=True)
```

**Baseline**

Omits KV-cache, ALiBi. RoPe added later. Minimal args, no forward loop branching.

```python
class Attention(nn.Module):
  def __init__(self, hidden_dim, num_heads, **kwargs):
    super().__init__()
    factory_kwargs = nn.factory_kwargs(kwargs)
    self.num_heads = num_heads    
    self.head_dim = hidden_dim // num_heads

    self.Q = nn.Linear(hidden_dim, num_heads * self.head_dim, bias=False, **factory_kwargs)
    self.K = nn.Linear(hidden_dim, num_heads * self.head_dim, bias=False, **factory_kwargs)
    self.V = nn.Linear(hidden_dim, num_heads * self.head_dim, bias=False, **factory_kwargs)
    self.proj = nn.Linear(hidden_dim, hidden_dim, bias=False, **factory_kwargs)


  def forward(self, x):
    B, T, _ = x.size()  
    q = self.Q(x).view(B, T, self.num_heads, self.head_dim).transpose(1, 2)
    k = self.K(x).view(B, T, self.num_heads, self.head_dim).transpose(1, 2)
    v = self.V(x).view(B, T, self.num_heads, self.head_dim).transpose(1, 2)
    
    attn = _attention(q, k, v)
    return self.proj(attn.transpose(1, 2).contiguous().view(B, T, -1))
```
Note: Norm (or RMS Norm) can be treated like activation functions and kept in the global function registry. GQA stays the same as here: https://github.com/ZHZisZZ/dllm/blob/871efe9dc578b04ec6d8d93a463907aa1d46930b/dllm/pipelines/llada/models/modeling_llada.py

### 3. Transformer

Minimal baseline without norm: 4 args plus activation (3 if dropout defaults), avoiding config. Alternatively, pass config dict at transformer block level and link to `transformer.pretrained.config`.

```python
class TF(nn.Module):
  def __init__(self, hidden_dim, num_heads, exp_factor, dropout_p, act_fn, **kwargs):
    super().__init__()
    factory_kwargs = nn.factory_kwargs(kwargs)
    
    self.attention = Attention(hidden_dim, num_heads, **factory_kwargs)
    self.mlp = MLP(hidden_dim, exp_factor, dropout_p, act_fn, **factory_kwargs)

  def forward(self, x):
    x = x + self.attention(x)
    x = x + self.mlp(x)
    return x
```

### **Outlook: Language model**

Omitted for simplicity: embedding layer, layer stacking, and LM head (standard boilerplate). Next step: add layer_id, align setup to load pretrained weights from target architecture (LLaDA, BERT, etc.).

**Open question:** Should composite layers implement `reset_parameters()` for activation-aware initialization (`calculate_gain`, `xavier_uniform_`) rather than relying on global model init? [PyTorch docs](https://docs.pytorch.org/docs/stable/nn.init.html)

```python
# Example pattern:
def reset_parameters(self):
  gain = nn.init.calculate_gain('relu')
  nn.init.xavier_uniform_(self.Q.weight, gain=gain)
  nn.init.xavier_uniform_(self.K.weight, gain=gain)
  nn.init.xavier_uniform_(self.V.weight, gain=gain)
  nn.init.xavier_uniform_(self.proj.weight, gain=gain)
```
